data_integration/logging/slack.py

```diff
@@ -1,71 +1,41 @@
-                self.node_output[key] = {True: [], False: []}
-
-            self.node_output[key][event.is_error].append(event)
-
-
-        elif isinstance(event, events.NodeFinished):
-            key = tuple(event.node_path)
-            if not event.succeeded and event.is_pipeline is False:
-
-                message = {'text': '\n:baby_chick: Ooops, a hiccup in '
-                                   + '_ <' + config.base_url() + '/' + '/'.join(event.node_path)
-                                   + ' | ' + '/'.join(event.node_path) + ' > _',
-                           'attachments': []}
-
-                if (self.node_output[key][False]):
-                    message['attachments'].append({'text': self.format_output(self.node_output[key][False]),
-                                                   'mrkdwn_in': ['text']})
-
-                if (self.node_output[key][True]):
-                    message['attachments'].append({'text': self.format_output(self.node_output[key][True]),
-                                                   'color': '#eb4d5c', 'mrkdwn_in': ['text']})
-
-                response = requests.post('https://hooks.slack.com/services/' + config.slack_token(), json=message)
-
-                if response.status_code != 200:
-                    raise ValueError(
-                        'Request to slack returned an error %s. The response is:\n%s' % (
-                            response.status_code, response.text)
-                    )
-
-    def format_output(self, output_events: [events.Output]):
-        output, last_format = '', ''
-        for event in output_events:
-            if event.format == events.Output.Format.VERBATIM:
-                if last_format == event.format:
-                    # append new verbatim line to the already initialized verbatim output
-                    output = output[0:-3] + '\n' + event.message + '```'
-                else:
-                    output += '\n' + '```' + event.message + '```'
-            elif event.format == events.Output.Format.ITALICS:
-                for line in event.message.splitlines():
-                    output += '\n _ ' + str(line) + ' _ '
-            else:
-                output = '\n' + event.message
-
-            last_format = event.format
-        return output
+from ..logging.chat_room import ChatRoom
+import requests
+import os
+
+
+class Slack(ChatRoom):
+    def __init__(self):
+        super().__init__(chat_type="Slack", code_markup_start="```", code_markup_end="```",
+                         line_start='\n _', line_end=' _ ')
+
+    def create_error_text(self, node_path: []):
+        path = '/'.join(node_path)
+        text = '\n:baby_chick: Ooops, a hiccup in ' + '_ <' + config.base_url() + '/' + path \
+               + ' | ' + path + ' > _'
+        return text
+
+    def create_error_msg(self, text, error_log1, error_log2):
+        message = {'text': text}
+        attachments = [{'text': error_log1, 'mrkdwn_in': ['text']},
```

Comment:
These have to be added conditionally: both loigs canbe empty

Reply:
Please tell me when it can be empty because in old code there are no conditions to check that these are empty or not.

Reply:
The old coffee guards against this case:

```py
                if (self.node_output[key][False]):
                    message['attachments'].append({'text': self.format_output(self.node_output[key][False]),
                                                   'mrkdwn_in': ['text']})

                 if (self.node_output[key][True]):
                    message['attachments'].append({'text': self.format_output(self.node_output[key][True]),
                                                   'color': '#eb4d5c', 'mrkdwn_in': ['text']})
```

Case for the non-error log could be empty: a task with a sql file with an error in the first line (select 1/0;). For no error log: a python program which just exits with non-success code?

Reply:
Yes, you are right.